src/apps/service/apps/job_listings.clj

```diff
@@ -68,18 +72,31 @@
     :batch           (:is_batch job)
     :batch_status    (when (:is_batch job) (format-batch-status id))}))
 
+(defn- interactive-urls
+  [job rep-steps]
+  (let [user-id (get-valid-user-id (:username job))
+        job-id  (:id job)
+        steps   (rep-steps job-id)]
+    (reduce (fn [urls step]
+              (when (= (:job_type step) jp/interactive-job-type)
+                (conj urls (str (interapps-url (url (config/interapps-base)) user-id (:external_id step))))))
```

Comment:
I don't think this will be a problem because all of our interactive apps currently only have one step, but I thought I'd mention it just in case. The reduction function currently clears the list of URLs whenever it encounters a non-interactive job step, so this function will always return only the consecutive interactive URLs at the end of the list of job steps. To get a better idea of what's happening, you can take a quick look at this testing I did in the REPL:

```
user=> (reduce (fn [acc n] (when (even? n) (conj acc n))) [] [1 2 3 4 5])
nil
user=> (reduce (fn [acc n] (when (even? n) (conj acc n))) [] [1 2 3 4])
(4)
user=> (reduce (fn [acc n] (when (even? n) (conj acc n))) [] [2 4 6 8])
[2 4 6 8]
user=> (reduce (fn [acc n] (when (even? n) (conj acc n))) [] [1 2 3 4 6 8])
(8 6 4)
user=> (reduce (fn [acc n] (when (even? n) (conj acc n))) '() '(1 2 3 4 6 8))
(8 6 4)
user=> (reduce (fn [acc n] (if (even? n) (conj acc n) acc)) [] [1 2 3 4 6 8])
[2 4 6 8]
```

You can fix the reduction function by having it return the list of URLs if the step that it's currently processing is not interactive:

```clojure
(defn- interactive-urls
  [job rep-steps]
  (let [user-id (get-valid-user-id (:username job))
        job-id  (:id job)
        steps   (rep-steps job-id)]
    (reduce (fn [urls step]
              (if (= (:job_type step) jp/interactive-job-type)
                (conj urls (str (interapps-url (url (config/interapps-base)) user-id (:external_id step))))
                urls))
            []
            steps)))
```

If you wanted to get fancy, you could use `map` and `filter` instead:

```clojure
(defn- interactive-urls
  [job rep-steps]
  (let [user-id      (get-valid-user-id (:username job))
        interactive? (fn [step] (= (:job-type step) jp/interactive-job-type))
        get-url      (fn [step] (str (interapps-url (url (config/interapps-base)) user-id (external_id step))))]
    (map get-url (filter interactive? (rep-steps (:id job))))))
```

I defined a couple of local functions just for readability, but that's completely optional.

The usual caveat that my example code hasn't been tested applies. 😄

Reply:
Good catch. I overlooked the `when` inside the reducer.
I think you'd typically want the reducer to always return the results so far, but `when` returns `nil` if the condition fails, so that's basically discarding the results-so-far when the check fails.

I think the `if` check makes more sense in the `reduce`, but I also like the `map` + `filter` solution too.

Reply:
Ooh, goodies to try!  I did not know that about reduce - good catch. I did try my own filter + map combo, but then I noticed for completely non-interactive jobs, I ended up with the `interactive_urls` key with an empty array and I wasn't sure if that was ideal. In my (broken) solution, the key wasn't there.  That being said, I haven't tried your fancy solution, and I like fancy - I'll give it a go!

Reply:
Goodies to try are always fun, and Clojure provides a lot of them. I love that about Clojure. :)

Both of my suggestions will result in an empty array under `:interactive_urls`, but you can fix that using `seq`, which converts any collection into a sequence and any empty collection into nil:

```clojure
user=> (seq [])
nil
user=> (seq [1 2 3])
(1 2 3)
user=> (seq '())
nil
user=> (seq {})
nil
user=> (seq {:foo :bar})
([:foo :bar])
user=> (seq '(1 2 3))
(1 2 3)
```

Probably the easiest option would be to put `seq` around the call to `reduce` or `map` depending on which option you choose.

Reply:
🌟 Fancy + seq works!  Thanks for taking the time to explain it all!  I'll push the changes shortly